File: src/main/kotlin/mergesort/Mergesort.java
package mergesort;

public final class Mergesort {
    public static void intArray(int[] array) {
        mergesort(array, 0, array.length - 1);
    }

    private static void mergesort(int[] array, int start, int end) {
        if (start < end) {
            int divisionIndex = (start + end) / 2;
            mergesort(array, start, divisionIndex);
            mergesort(array, divisionIndex + 1, end);
            merge(array, start, divisionIndex, end);
        }
    }

    private static void merge(int[] array, int start, int middle, int end) {
        int leftArrayLength = middle - start + 1;
        int rightArrayLength = end - middle;
        int[] leftArray = new int[leftArrayLength];
        int[] rightArray = new int[rightArrayLength];
        System.arraycopy(array, start, leftArray, 0, leftArrayLength);
        System.arraycopy(array, middle + 1, rightArray, 0, rightArrayLength);
        int arrayIndex = start;
        int leftIndex = 0;
        int rightIndex = 0;
        while (leftIndex < leftArray.length || rightIndex < rightArray.length) {
            if (leftIndex >= leftArray.length) {
                array[arrayIndex++] = rightArray[rightIndex++];
            } else if (rightIndex >= rightArray.length) {
                array[arrayIndex++] = leftArray[leftIndex++];
            } else {
                if (rightArray[rightIndex] < leftArray[leftIndex]) {
                    array[arrayIndex++] = rightArray[rightIndex++];
                } else {
                    array[arrayIndex++] = leftArray[leftIndex++];
                }
            }
        }
    }
}

File: src/main/kotlin/mergesort/Mergesort.kt
package mergesort

internal object Mergesort {
    fun intArray(array: IntArray) {
        if (array.size >= 2)
            mergesort(array, 0, array.size - 1)
    }

    private fun mergesort(array: IntArray, start: Int, end: Int) {
        if (start < end) {
            val divisionIndex: Int = start + (end - start) / 2
            mergesort(array, start, divisionIndex)
            mergesort(array, divisionIndex + 1, end)
            merge(array, start, divisionIndex, end)
        }
    }

    private fun merge(array: IntArray, firstIndex: Int, middleIndex: Int, lastIndex: Int) {
        val arrayOne = array.slice(firstIndex..middleIndex)
        val arrayTwo = array.slice(middleIndex + 1..lastIndex)
        val ret = IntArray(lastIndex - firstIndex + 1)
        var i = 0
        var indexOne = 0
        var indexTwo = 0
        while (indexOne < arrayOne.size || indexTwo < arrayTwo.size) {
            if (indexOne >= arrayOne.size) {
                ret[i] = arrayTwo[indexTwo++]
            } else if (indexTwo >= arrayTwo.size) {
                ret[i] = arrayOne[indexOne++]
            } else {
                if (arrayOne[indexOne] <= arrayTwo[indexTwo]) {
                    ret[i] = arrayOne[indexOne++]
                } else {
                    ret[i] = arrayTwo[indexTwo++]
                }
            }
            i++
        }
        ret.forEachIndexed { index, element ->
            array[index + firstIndex] = element
        }
    }
}
